src/firebase_functions/firestore_fn.py

```diff
@@ -146,34 +172,49 @@ def _firestore_endpoint_handler(
             firestore_event_data.old_value.update_time,
         )
     if event_type == _event_type_deleted:
-        firestore_event_data = old_value_snapshot
+        firestore_event_data = _typing.cast(_firestore.DocumentEventData,
+                                            old_value_snapshot)
     if event_type == _event_type_created:
-        firestore_event_data = value_snapshot
+        firestore_event_data = _typing.cast(_firestore.DocumentEventData,
+                                            value_snapshot)
     if event_type in (_event_type_written, _event_type_updated):
-        firestore_event_data = Change(
-            before=old_value_snapshot,
-            after=value_snapshot,
-        )
+        firestore_event_data = _typing.cast(
+            _firestore.DocumentEventData,
+            Change(
+                before=old_value_snapshot,
+                after=value_snapshot,
+            ))
 
     params: dict[str, str] = {
         **document_pattern.extract_matches(event_document),
     }
-    database_event = Event(
-        project=event_project,
-        namespace=event_namespace,
-        database=event_database,
-        location=event_location,
-        document=event_document,
-        specversion=event_attributes["specversion"],
-        id=event_attributes["id"],
-        source=event_attributes["source"],
-        type=event_attributes["type"],
-        time=event_time,
-        data=firestore_event_data,
-        subject=event_attributes["subject"],
-        params=params,
-    )
-    func(database_event)
+
+    common_event_kwargs = {
+        "project": event_project,
+        "namespace": event_namespace,
+        "database": event_database,
+        "location": event_location,
+        "document": event_document,
+        "specversion": event_attributes["specversion"],
+        "id": event_attributes["id"],
+        "source": event_attributes["source"],
+        "type": event_attributes["type"],
+        "time": event_time,
+        "data": firestore_event_data,
+        "subject": event_attributes["subject"],
+        "params": params,
+    }
+
+    if event_type.endswith(".withAuthContext"):
+        database_event_with_auth_context = EventWithAuthContext(
+            **common_event_kwargs,
+            auth_type=event_auth_type,
+            auth_id=event_auth_id)
+        func(database_event_with_auth_context)
+    else:
+        database_event = Event(**common_event_kwargs)
+        # mypy cannot infer that the event type is correct, hence the cast
+        _typing.cast(_C1 | _C2, func)(database_event)
```

Comment:
Can probably keep this as an `Event` type, then if it's auth context event use `__dict__ ` to append it to `EventWithAuthContext`, something like:
```suggestion
   database_event = Event(
        project=event_project,
        namespace=event_namespace,
        database=event_database,
        location=event_location,
        document=event_document,
        specversion=event_attributes["specversion"],
        id=event_attributes["id"],
        source=event_attributes["source"],
        type=event_attributes["type"],
        time=event_time,
        data=firestore_event_data,
        subject=event_attributes["subject"],
        params=params,
    )

    if event_type.endswith(".withAuthContext"):
        database_event_with_auth_context = EventWithAuthContext(
            **vars(database_event),
            auth_type=event_auth_type,
            auth_id=event_auth_id)
        func(database_event_with_auth_context)
    else:
        func(database_event)
```